### DISCUSSION TITLE
can't access request's body inside custom exception handler for logging

### DISCUSSION THREAD

**Discussion by naveenmarthala** [fastapi/fastapi#6509] [Questions] [answered]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Description

I am trying to access the body of the request for logging purposes inside the function, in which I am handling the payloads that raise `RequestValidationError`.

### Example Code
this is the function, which gets triggered on `RequestValidationError`s and inside which I am trying to log the body of the requests that raise the error to a log file.
```python
from fastapi import Request
from fastapi.exceptions import RequestValidationError, ValidationError
from fastapi.responses import JSONResponse

@app.exception_handler(RequestValidationError)
@app.exception_handler(ValidationError)
async def invalid_input_handler(request: Request, exc:Union[RequestValidationError, ValidationError]) -> Dict[str, List[str]]:
    # get the body to a sring. 
    body = await request.body()
    # log all the relevant things
    error_logger.error(f'{request.url} - body: {body} - error: {exc}')
    # but, send a pre-made response to the client.
    return JSONResponse(
        status_code=422,
        content={
            'predictions': '-999.99',
            'message': '<my-message>'
            }
        )
```

1. `await request.body()` never completes the request and kept the client waiting indefinetely. so, disconnecting from client prints nothing to my error-logger that i am using to track errors, but prints this to gunicorn logger with DEBUG level:

```
Traceback (most recent call last):
  File "/usr/local/lib/python3.8/site-packages/fastapi/routing.py", line 205, in app
    json_body = await request.json()
  File "/usr/local/lib/python3.8/site-packages/starlette/requests.py", line 235, in json
    self._json = json.loads(body)
  File "/usr/local/lib/python3.8/json/__init__.py", line 357, in loads
    return _default_decoder.decode(s)
  File "/usr/local/lib/python3.8/json/decoder.py", line 337, in decode
    obj, end = self.raw_decode(s, idx=_w(s, 0).end())
  File "/usr/local/lib/python3.8/json/decoder.py", line 353, in raw_decode
    obj, end = self.scan_once(s, idx)
json.decoder.JSONDecodeError: Expecting ',' delimiter: line 1 column 34 (char 33)

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/local/lib/python3.8/site-packages/starlette/exceptions.py", line 71, in __call__
    await self.app(scope, receive, sender)
  File "/usr/local/lib/python3.8/site-packages/starlette/routing.py", line 656, in __call__
    await route.handle(scope, receive, send)
  File "/usr/local/lib/python3.8/site-packages/starlette/routing.py", line 259, in handle
    await self.app(scope, receive, send)
  File "/usr/local/lib/python3.8/site-packages/starlette/routing.py", line 61, in app
    response = await func(request)
  File "/usr/local/lib/python3.8/site-packages/fastapi/routing.py", line 211, in app
    raise RequestValidationError([ErrorWrapper(e, ("body", e.pos))], body=e.doc)
fastapi.exceptions.RequestValidationError: 1 validation error for Request
body -> 33
  Expecting ',' delimiter: line 1 column 34 (char 33) (type=value_error.jsondecode; msg=Expecting ',' delimiter; doc={"input_json" : [6.3, 2.9, 4.9, 1.]}; pos=33; lineno=1; colno=34)

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/local/lib/python3.8/site-packages/uvicorn/protocols/http/httptools_impl.py", line 378, in run_asgi
    result = await app(self.scope, self.receive, self.send)
  File "/usr/local/lib/python3.8/site-packages/fastapi/applications.py", line 212, in __call__
    await super().__call__(scope, receive, send)
  File "/usr/local/lib/python3.8/site-packages/starlette/applications.py", line 112, in __call__
    await self.middleware_stack(scope, receive, send)
  File "/usr/local/lib/python3.8/site-packages/starlette/middleware/errors.py", line 181, in __call__
    raise exc
  File "/usr/local/lib/python3.8/site-packages/starlette/middleware/errors.py", line 159, in __call__
    await self.app(scope, receive, _send)
  File "/usr/local/lib/python3.8/site-packages/starlette/exceptions.py", line 90, in __call__
    response = await handler(request, exc)
  File "/home/dockerd/api/main.py", line 79, in invalid_input_handler
    body = await request.body()
  File "/usr/local/lib/python3.8/site-packages/starlette/requests.py", line 227, in body
    async for chunk in self.stream():
  File "/usr/local/lib/python3.8/site-packages/starlette/requests.py", line 221, in stream
    raise ClientDisconnect()
starlette.requests.ClientDisconnect
```

2. while using `request.body()`:
    for input request: `{"input_json" : [6.3, 2.9, 4.9, 1.]}`
    this gets added to the logs(check above code to see how i am logging this):
    > [2022-01-29 07:30:32 +0000] [ERROR] - http://0.0.0.0:8000/predict_class - body: <coroutine object Request.body at 0x7f6dcea711c0> - error: 1 validation error for Request
    > body -> 33
    >   Expecting ',' delimiter: line 1 column 34 (char 33) (type=value_error.jsondecode; msg=Expecting ',' delimiter; doc={"input_json" : [6.3, 2.9, 4.9, 1.]}; pos=33; lineno=1; colno=34)

    and this to gunicorn-logger with 'DEBUG' log level, which i presume was printed to std.err:

    > /usr/local/lib/python3.8/site-packages/starlette/exceptions.py:90: RuntimeWarning: coroutine 'Request.body' was never awaited
    >   response = await handler(request, exc)
    > RuntimeWarning: Enable tracemalloc to get the object allocation traceback

### Operating System

Linux

### Operating System Details

Debian 11 (docker image tagged python:3.8-slim)

### FastAPI Version
0.73.0
versions of other relevant packages:
```
$ pip list | grep -E 'fastapi|uvicorn|gunicorn|starlette|pydantic'
fastapi           0.73.0
gunicorn          20.1.0
pydantic          1.9.0
starlette         0.17.1
uvicorn           0.17.1
```

### Python Version

3.8.12 (docker image tagged python:3.8-slim)

### Additional Context

`gunicorn api.main:app --worker-class 'uvicorn.workers.UvicornWorker' --user dockerd --capture-output --keep-alive 0 --log-level debug --log-file /home/dockerd/gunicorn_logs/gunicorn_info.log` is how I am running fastapi using gunicorn and uvicorn.

How do I access the `request.body()` as a raw string to use in my log messages?

**Comment by mdtomo** [comment, marked as answer]
You can access the body directly from the exception (`exc.body`). See this link to the [docs](https://fastapi.tiangolo.com/tutorial/handling-errors/#use-the-requestvalidationerror-body).
